crates/iota-data-ingestion-core/README.md

```diff
@@ -0,0 +1,352 @@
+
+```rust
+pub async fn register<W: Worker + 'static>(&mut self, pool: WorkerPool<W>) -> Result<()> {..}
+```
+
+After instantiation, the `IndexerExecutor` requires registration of one or more `WorkerPool` instances via the `register(..)` method to function. During registration, the executor retrieves the last recorded watermark from its `ProgressStoreWrapper`, using it as the initial checkpoint sequence number for synchronization. The `register` method also creates a buffered `mpsc` channel with a capacity of `MAX_CHECKPOINTS_IN_PROGRESS` and stores the `WorkerPool` in its internal `pools` vector. While the `pool.run()` method is called, it does not begin execution until awaited. The sender half of the created channel is stored in `pool_senders` vector for later distribution of `CheckpointData`.
+
+```rust
+pub async fn run(
+    mut self,
+    // The directory on which checkpoints will be read
+    path: PathBuf,
+    // Provide a remote URL to download checkpoints
+    remote_store_url: Option<String>,
+    // Any config key-value pair for the Remote fetcher
+    remote_store_options: Vec<(String, String)>,
+    // Customize and optimize the Checkpoint fetch behavior
+    reader_options: ReaderOptions,
+    // Graceful shutdown signal
+    mut exit_receiver: oneshot::Receiver<()>,
+) -> Result<ExecutorProgress> {..}
+```
+
+The `run(..)` method executes the actor's main logic. The `run(..)` method is the main orchestrator of the checkpoint processing pipeline. It coordinates between:
+
+- Checkpoint reading (local/remote)
+- Worker pool management
+- Progress tracking
+- Garbage collection
+- Metric reporting
+
+The following diagram shows on a high level the overall flow:
+
+```mermaid
+sequenceDiagram
+    participant CR as CheckpointReader
+    participant IE as IndexerExecutor
+    participant WP as WorkerPool
+    participant PS as ProgressStore
+
+    IE->>CR: Initialize
+    IE->>WP: Spawn Workers
+
+    loop Until Exit
+        CR->>IE: New Checkpoint (checkpoint_recv)
+        IE->>WP: Distribute (pool_senders)
+
+        Note over WP: Process Checkpoint
+
+        WP->>IE: Progress Update (pool_progress_sender)
+        IE->>PS: Save Progress
+        PS->>IE: Return min_watermark
+
+        alt Watermark Advanced
+            IE->>CR: GC Signal (gc_sender)
+        end
+    end
+```
+
+To initiate checkpoint fetching, the `CheckpointReader` requires an initial checkpoint sequence number. This number is obtained by calling the `ProgressStoreWrapper`'s `min_watermark()` method, which determines the minimum processed checkpoint sequence number across all worker pools. This ensures that no checkpoints are skipped during synchronization and serves as the garbage collection starting point. Subsequently, the `CheckpointReader` is started in a separate task by invoking its `run()` method.
```

Comment:
Another place where links to methods would be helpful.

Reply:
[fixed](https://github.com/iotaledger/iota/pull/4570/commits/c58ed8cab8960cc11588dd32e7a540ae8ebe4c3c#diff-31a167158fe911c55387e5f89c28edd3fd006968b761a550e15148aee28192cbR309)